paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/changelog/ChangelogCompactOperator.java

```diff
@@ -0,0 +1,299 @@
+                    message.partition(),
+                    message.bucket(),
+                    false,
+                    meta);
+        }
+        for (DataFileMeta meta : message.compactIncrement().changelogFiles()) {
+            copyFile(
+                    dataFilePathFactory.toPath(meta.fileName()),
+                    message.partition(),
+                    message.bucket(),
+                    true,
+                    meta);
+        }
+
+        // send commit message without changelog files
+        CommitMessageImpl newMessage =
+                new CommitMessageImpl(
+                        message.partition(),
+                        message.bucket(),
+                        new DataIncrement(
+                                message.newFilesIncrement().newFiles(),
+                                message.newFilesIncrement().deletedFiles(),
+                                Collections.emptyList()),
+                        new CompactIncrement(
+                                message.compactIncrement().compactBefore(),
+                                message.compactIncrement().compactAfter(),
+                                Collections.emptyList()),
+                        message.indexIncrement());
+        Committable newCommittable =
+                new Committable(committable.checkpointId(), Committable.Kind.FILE, newMessage);
+        if (record.hasTimestamp()) {
+            output.collect(new StreamRecord<>(newCommittable, record.getTimestamp()));
+        } else {
+            output.collect(new StreamRecord<>(newCommittable));
+        }
+    }
+
+    private void copyFile(
+            Path path, BinaryRow partition, int bucket, boolean isCompactResult, DataFileMeta meta)
+            throws Exception {
+        if (!outputStreams.containsKey(partition)) {
+            Path outputPath =
+                    new Path(path.getParent(), "tmp-compacted-changelog-" + UUID.randomUUID());
+            outputStreams.put(
+                    partition,
+                    new OutputStream(
+                            outputPath, table.fileIO().newOutputStream(outputPath, false)));
+        }
+
+        OutputStream outputStream = outputStreams.get(partition);
+        long offset = outputStream.out.getPos();
+        try (SeekableInputStream in = table.fileIO().newInputStream(path)) {
+            IOUtils.copyBytes(in, outputStream.out, IOUtils.BLOCKSIZE, false);
+        }
+        table.fileIO().deleteQuietly(path);
+        results.computeIfAbsent(partition, p -> new ArrayList<>())
+                .add(
+                        new Result(
+                                bucket,
+                                isCompactResult,
+                                meta,
+                                offset,
+                                outputStream.out.getPos() - offset));
+
+        if (outputStream.out.getPos() >= table.coreOptions().targetFileSize(false)) {
```

Comment:
Only pk table has changelog, so why here call targetFileSize with false?

Reply:
Changelog files are not LSM tree data files. They're just a bunch of records and they don't care about what the primary keys are.

Reply:
OK, you are right.